### PR TITLE
Demo mode turret patch

### PR DESCRIPTION
**Pull request by nicholasl23638**
Turret class in main working and we need a demo mode working before Thur

Patches:
Implemented a working demo mode option in yml controlboard

Changes to turret:
1. Removed zeroOffset logic - we aren't masking anymore (neither % nor &) and turret is now able to handle negative values - removed zeroOffset to facilitate comparison desired/actual turret positions
2. Simplified turret zeroSensors - no longer fiddling with zeroOffset, so we no longer need to set turretMotor pos on bootup - removed abs encoder zeroing b/c the whole point of an abs encoder is it keeps its pos - removed ratio checking b/c we should only boot up turret position using abs enc if turretPPR:absPPR = 1

Note: Patches work in simulation but have yet to be tested irl

### CODE REVIEW

File: src/main/java/com/team1816/season/subsystems/Turret.java [hunk 1]
```diff
@@ -61,107 +59,92 @@ public class Turret extends Subsystem implements PidProvider {
...
+        zeroSensors(false);
+    }
+
+    public synchronized void zeroSensors(boolean resetEncPos) {
+        desiredPos = 0;
+        followingPos = 0;
+        lostEncPos = false;
+
+        if (resetEncPos) {
+            turretMotor.setSelectedSensorPosition(
```
Comment (richiksc) [inline src/main/java/com/team1816/season/subsystems/Turret.java:143]: Unless I misunderstood new logic, this will break the turret - this sets the selectedSensorPosition (relative encoder position) to 0. What the previous logic did was set it to the absolute encoder position. It rehomed the the quadrature (relative) encoder to the source of truth absolute (pulse width) encoder.

The absolute encoder was never being zeroed.

File: src/main/java/com/team1816/season/subsystems/Turret.java [hunk 2]
```diff
@@ -411,23 +368,28 @@ private void trackAbsolute() {
...
-            int adjPos = (rawPos + ABS_TICKS_SOUTH + ZERO_OFFSET);
 
-            turret.set(com.ctre.phoenix.motorcontrol.ControlMode.Position, adjPos);
-            outputsChanged = false;
+            if (pos > kFwdWrapAroundPos) {
+                pos -= kTurretPPR;
+            } else if (pos < kRevWrapAroundPos) {
+                pos += kTurretPPR;
+            }
+            int rawPos = (pos - kAbsTicksSouthOffset);
```
Comment (richiksc) [inline src/main/java/com/team1816/season/subsystems/Turret.java:384]: Should this be plus instead of minus?

```suggestion
            int rawPos = (pos + kAbsTicksSouthOffset);
```

File: src/main/java/com/team1816/season/subsystems/Turret.java [hunk 3]
```diff
@@ -172,37 +155,11 @@ public ControlMode getControlMode() {
     public void setControlMode(ControlMode controlMode) {
         if (this.controlMode != controlMode) {
             outputsChanged = true;
-            if (
-                controlMode == ControlMode.CAMERA_FOLLOWING ||
-                controlMode == ControlMode.CAMERA_SNAP
-            ) {
-                if (Constants.kUseVision) {
```
Comment (richiksc) [inline src/main/java/com/team1816/season/subsystems/Turret.java:179]: Where is the control mode logic handled now?

File: src/main/java/com/team1816/season/Robot.java [hunk 4]
```diff
@@ -531,6 +543,11 @@ public void disabledPeriodic() {
                         autoModeManager.getSelectedAuto().getCurrentTrajectory()
                     );
             }
+
+            // check if demo mode speed multiplier changed
+            if (ControlBoardBrige.getInstance().isDemoMode()) {
+                controlBoard.update();
```
Comment (richiksc) [inline src/main/java/com/team1816/season/Robot.java:549]: Should be a better way to do this than polling in disabledPeriodic. What was the issue with the entry listener?

**Review by richiksc — CHANGES_REQUESTED** [review]

File: src/main/java/com/team1816/season/Robot.java [hunk 5]
```diff
@@ -280,11 +280,15 @@ public void robotInit() {
                     // Operator Gamepad
                     createAction(
                         () -> controlBoard.getAsBool("raiseBucket"),
-                        () -> distanceManager.incrementBucket(100)
+                        () -> {
```
Comment (richiksc) [inline src/main/java/com/team1816/season/Robot.java:283]: undo this change and line 289-291 change?

Comment (richiksc) [inline src/main/java/com/team1816/season/Robot.java:283]: undo this change and line 289-291 change?

File: src/main/java/com/team1816/season/states/Superstructure.java [hunk 6]
```diff
@@ -125,6 +125,7 @@ public void setFiring(boolean firing) {
     public void autoAim() {
         if (Constants.kUseVision) {
             ledManager.setCameraLed(true);
+            camera.setCameraEnabled(true);
```
Comment (richiksc) [inline src/main/java/com/team1816/season/states/Superstructure.java:128]: good

File: src/main/java/com/team1816/season/subsystems/Turret.java [hunk 7]
```diff
@@ -81,15 +82,15 @@ public Turret() {
 
         // define limits + when turret should wrap around
         kRevLimit =
-            Math.min(
-                (int) factory.getConstant(NAME, "fwdLimit"),
-                ((int) factory.getConstant(NAME, "revLimit"))
-            );
+            //            Math.min(
```
Comment (richiksc) [inline src/main/java/com/team1816/season/subsystems/Turret.java:85]: remove commented out code?

**Review by richiksc — APPROVED** [review]

File: src/main/java/com/team1816/season/Robot.java (hunk 4 above)
Comment (ddaymn) [reply src/main/java/com/team1816/season/Robot.java:557]: Yes this should be event driven.

File: src/main/java/com/team1816/season/Robot.java [hunk 8]
```diff
@@ -596,7 +621,7 @@ public void manualControl() {
         }
 
         drive.setTeleopInputs(
-            controlBoard.getAsDouble("throttle"),
+            -controlBoard.getAsDouble("throttle"),
```
Comment (ddaymn) [inline src/main/java/com/team1816/season/Robot.java:624]: This is ok for now but ultimately; we should not have to do this.  We know if motors are inverted etc that should drive the need for this

File: src/main/java/com/team1816/season/controlboard/ControlBoard.java [hunk 9]
```diff
@@ -27,37 +29,14 @@ private ControlBoard(Controller.Factory controller) {
...
+            SmartDashboard.putData("Demo Mode", demoModeChooser);
             for (DemoMode demoMode : DemoMode.values()) {
                 demoModeChooser.addOption(demoMode.name(), demoMode);
             }
             demoModeChooser.setDefaultOption(DemoMode.SLOW.name(), DemoMode.SLOW);
-            NetworkTableInstance
-                .getDefault()
-                .getTable("SmartDashboard")
-                .getSubTable("DemoMode")
-                .addEntryListener(
```
Comment (ddaymn) [inline src/main/java/com/team1816/season/controlboard/ControlBoard.java:40]: the listner is how you should be getting the change rather than the disabled check

File: src/main/java/com/team1816/season/Robot.java (hunk 8 above)
Comment (richiksc) [reply src/main/java/com/team1816/season/Robot.java:624]: Y axis in general should be negated, input from controller is opposite of logical. This is why the negation is applied right now. Motor inversion is fine, positive input to drive results in forward motion.
